Add `Hrp` type

tcharding opened the pull request:
Instead of accepting `&str` for hrp strings we can add a type `Hrp` and abstract the checking of the string. This is the "parse don't validate" methodology suggested by clarkmoody [here](https://github.com/rust-bitcoin/rust-bech32/pull/100#issuecomment-1512435732).

This is an invasive breaking change.

tcharding commented:
Please note, I'm not happy with the inefficient way this PR converts an `Hrp` to a lowercase version in `encode_to_fmt`.

clarkmoody commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,180 @@
+
+    /// Creates a lowercase iterator over the bytes (ASCII characters) of this HRP.
+    pub fn iter(&self) -> LowercaseIter<'s> {
+        let s = match self {
+            Self::Upper(s) => s,
+            Self::Lower(s) => s,
+            Self::Digit(s) => s,
+        };
```
Reuse `self.as_str` here and in `len` below?

tcharding replied:
Works for length but I get lifetime errors here. Thanks! I'll give you time to finish review before I force push.

clarkmoody replied:
This seems to work for me (put `'s` lifetime on `self`)

```rust
    pub fn iter(&'s self) -> LowercaseIter<'s> {
        let iter = self.as_str().bytes();
        LowercaseIter { iter }
    }
```

tcharding replied:
haha thanks man, you can code for me anytime :)

tcharding replied:
Even better, we can just elide the lifetime on `LowercaseIter<'_>`

clarkmoody replied:
👌

apoelstra commented on src/lib.rs:
```diff
@@ -722,13 +674,11 @@ fn split_and_decode(s: &str) -> Result<(String, Vec<u5>), Error> {
             if c.is_lowercase() {
                 match case {
                     Case::Upper => return Err(Error::MixedCase),
-                    Case::None => case = Case::Lower,
```
In e0bfa91dc74888005246ae80b46d0a620827888f:

Removing these `Case::None` cases introduces a bug in case the HRP is all digits but the data part has uppercase letters.

tcharding replied:
Added this test, does this cover what you meant?
```rust
    
    #[test]
    #[cfg(feature = "alloc")]
    fn attempt_decode_all_digit_hrp_mixed_case_data() {
        let addr = "23451qar0sRRR7xfkvy5l643lydnw9re59gtzzwf5mdq";
        assert!(crate::decode(addr).is_err());
    }
```

tcharding replied:
I believe this is ok now, I added a test to verify it.
```rust
    #[test]
    #[cfg(feature = "alloc")]
    fn decode_all_digit_hrp_uppercase_data() {
        let addr = "23451QAR0SRRR7XFKVY5L643LYDNW9RE59GTZZLKULZK";
        let (hrp, data, variant) = crate::decode(&addr).expect("address is well formed");
        assert_eq!(hrp, Hrp::parse("2345").unwrap().0);
        let (hrp, _) = Hrp::parse("2345").unwrap();
        let s = crate::encode(hrp, &data, variant).expect("failed to encode");
        assert_eq!(s, addr);
    }
```

tcharding replied:
I added `Case::None` back in and added this test
```rust  
    #[test]
    #[cfg(feature = "alloc")]
    fn decode_all_digit_hrp_uppercase_data() {
        let addr = "23451QAR0SRRR7XFKVY5L643LYDNW9RE59GTZZLKULZK";
        let (hrp, data, variant) = crate::decode(&addr).expect("address is well formed");
        assert_eq!(hrp, Hrp::parse("2345").unwrap());
        let hrp = Hrp::parse("2345").unwrap();
        let s = crate::encode(hrp, &data, variant).expect("failed to encode");
        assert_eq!(s.to_uppercase(), addr);
    }
```

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,180 @@
+    /// Uppercase human-readable part.
+    Upper(&'a str),
+    /// All digit human-readable part.
+    Digit(&'a str),
```
I guess we will need it if we want to fix the case-checking bug in the top-level parsing function.... but I don't think we should bother, I think we should have a `check_characters` function that does validity checks on the whole string rather than checking for consistency with the HRP.

Once we do that I'm not sure that there's any value in having distinct `Lower`/`Upper` variants at all...we should just unconditionally use bitshifts to convert case when it's needed.

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,180 @@
+        match self {
+            Self::Upper(upper) => Cow::Owned(upper.to_lowercase()),
+            Self::Lower(lower) => Cow::Borrowed(lower),
+            Self::Digit(digits) => Cow::Borrowed(digits),
```
I guess this would be less efficient if we weren't tracking case, but I don't really see the use of this method anyway, so I think that's fine.

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,180 @@
+const MAX_HRP_LEN: usize = 83;
+
+/// The human-readable part (human readable prefix before the '1' separator).
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
```
In e0bfa91dc74888005246ae80b46d0a620827888f:

I think we should do a case-insensitive comparison rather than deriving PartialEq/Eq.

apoelstra replied:
And I think we should impl `Ord` similarly.

tcharding replied:
Done!

tcharding commented on src/lib.rs:
```diff
@@ -474,74 +477,32 @@ impl<U: AsRef<[u8]>> CheckBase32<()> for U {
-/// * If human-readable part (`hrp`) is invalid.
 /// * Deviations from standard.
 /// * No length limits are enforced for the data part.
 #[cfg(feature = "alloc")]
```
This is fixed now, no alloc feature needed for `encode_to_fmt`.

tcharding commented:
Changes in force push are substantial, `Hrp` now has an 83 byte long buffer which stores the string as ascii byte values. We provide iterators for getting bytes, chars, lowercase bytes, and lowercase chars.

Please note the logic in `split_and_decode` is changed, the returned `Hrp` has not been lowercased like the old string that was returned.

tcharding commented:
Converted to draft so we do not merge because `./fuzz.sh encode_decode` crashes on my machine with the following hex:
045f5f5f5f5f5f5f5f0000000000000000083535
ffa8531a004b3e0000b1f15ab026bd0d4fc2eb72825f7cf1570000000000
0000040000000000

Two interesting things:
1. I cannot reproduce the failures, this test passes

```rust

    #[test]
    fn duplicate_crash() {
        let mut a = Vec::new();
        extend_vec_from_hex("045f5f5f5f5f5f5f5f0000000000000000083535", &mut a);
        super::do_test(&a);

        let mut a = Vec::new();
        extend_vec_from_hex("ffa8531a004b3e0000b1f15ab026bd0d4fc2eb72825f7cf1570000000000", &mut a);
        super::do_test(&a);

        let mut a = Vec::new();
        extend_vec_from_hex("0000040000000000", &mut a);
        super::do_test(&a);
    }
```

2. The fuzzer always fails with the same values so it must be starting from the same initial state, this is what we were wondering about @apoelstra.

apoelstra commented:
@tcharding can you reproduce the failures if you test with `RUSTFLAGS=--cfg=fuzzing`?

If so you have been burned by the kind of bullshit that https://github.com/rust-bitcoin/rust-bitcoin/pull/1821 will fix.

apoelstra commented:
> 2\. The fuzzer always fails with the same values so it must be starting from the same initial state, this is what we were wondering about @apoelstra.

This does not sound like a valid inference.

tcharding commented:
> > 2. The fuzzer always fails with the same values so it must be starting from the same initial state, this is what we were wondering about @apoelstra.
> 
> This does not sound like a valid inference.

So I guessed from your comment that I don't understand how fuzzers work properly. I did some more reading. From my understanding the fuzzer starts with some input, runs the test, measures the coverage, then mutates the input in some interesting way and repeats. I thought that the initial input was random, hence why I was surprised that it failed with the same values on multiple runs.

Where is the hole in my understanding?

tcharding commented:
> @tcharding can you reproduce the failures if you test with `RUSTFLAGS=--cfg=fuzzing`?
> 
> If so you have been burned by the kind of bullshit that [rust-bitcoin/rust-bitcoin#1821](https://github.com/rust-bitcoin/rust-bitcoin/pull/1821) will fix.

I can't repro. Had a go this morning, left it a while, and had another go but still no dice?

apoelstra commented:
@tcharding try deleting your `hfuzz_workspace` directory then try again.

>Where is the hole in my understanding?

There are many paths to the same state.

But also, in this case, I think you're confused by the hex output from `./travis-fuzz.sh` which just looks in the `hfuzz_workspace` directory for the first crash file ... so if you don't delete that between runs, it'll output the same file. (And if you fix the bug, it may confusingly still be output.)

tcharding commented:
ooooh, would it be useful to add at the start of the loop (to all the fuzzing scripts):
```bash

# Clean up crash file from previous runs.
if [ -f hfuzz_workspace/$FILE/HONGGFUZZ.REPORT.TXT ]; then
        rm hfuzz_workspace/$FILE/HONGGFUZZ.REPORT.TXT
fi
```

tcharding commented:
Fixed up a few docs mistakes and removed infrastructure changes to fuzz files. @apoelstra please review carefully the changes in `encode_decode` - I feel like I'm still not groking fuzzing, I'm unsure if the code I've written is the right way to write fuzzing code.

tcharding commented:
Changes in force push, I found a bug while using this in `bitcoin`. Added a test that catches it and fixed it. (I wasn't using the lowercase iterator when writing to polymod function). Was not being caught because we lowercase in the API functions higher up.

apoelstra commented on fuzz/fuzz_targets/encode_decode.rs:
```diff
@@ -2,20 +2,33 @@ extern crate bech32;
 
-    let hrp_end = (data[0] as usize) + 1;
+    // Get a random valid length hrp 1-83 bytes.
+    let hrp_end = ((data[0] as usize) % 82) + 1;
```
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

This actually only gives numbers in the range 1-82.

tcharding replied:
ugh, code comments for the win!

apoelstra commented on fuzz/fuzz_targets/encode_decode.rs:
```diff
@@ -2,20 +2,33 @@ extern crate bech32;
+
+    let hrp = hrp.iter().take(hrp_end).map(|b| unsafe {
+        char::from_u32_unchecked(*b as u32)
+    }).collect::<String>();
```
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

Why is this safe?

apoelstra replied:
Oh, I see, you're translating the data in the lines above. I wouldn't do that.

I'd just take the first byte as a "length", try to parse up to that many bytes as an HRP, and use the rest as bech32 data.

apoelstra replied:
I'd also assert that the HRP round-trips.

tcharding replied:
> I'd just take the first byte as a "length", try to parse up to that many bytes as an HRP

So that's something I don't understand, even with the tight requirements on the ascii byte range will the fuzzer find data that parses?

apoelstra replied:
@tcharding yep. It's kinda magical. Basically it will see that whenever it goes out of bounds, this short-circuits the entire logic chain, so it shouldn't do that (or rather, after doing it once, there is nothing more for it to learn). So it is able to very quickly zero in on valid strings for pretty-much any definition of "valid".

tcharding replied:
Woke up this morning thinking about fuzzing, realized I could throw a panic in the fuzz test to see how long it took to hit (had not yet read your post at that stage), it hits it immediately - that is magic, what genius invented that methodology?

apoelstra commented on fuzz/fuzz_targets/encode_decode.rs:
```diff
@@ -28,11 +41,13 @@ fn do_test(data: &[u8]) {
-        let decoded = bech32::decode(&data_str);
-        let b32 = decoded.expect("should be able to decode own encoding");
+    if let Ok(hrp) = Hrp::parse(&hrp) {
+        if let Ok(encoded) = bech32::encode(hrp, &dp, variant).map(|b32| b32.to_string()) {
```
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

What is the purpose of this `.to_string`?

tcharding replied:
derp, what was I even mapping.

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,420 @@
+    /// > specific applications.
+    ///
+    /// [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+    pub fn parse(hrp: &str) -> Result<Self, Error> { Self::parse_and_case(hrp).map(|(p, _)| p) }
```
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

You can use `.0` as shorthand for `|(p, _)| p`.

tcharding replied:
double derp, was I on crack yesterday?

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,420 @@
+        for (i, b) in hrp.bytes().enumerate() {
+            // Valid subset of ASCII
+            if !(33..=126).contains(&b) {
+                return Err(InvalidAsciiByte(b));
```
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

It would probably be more user-friendly to return a character here instead of a byte.

tcharding replied:
I came to the conclusion that converting a byte that was pulled out of a unicode point was not useful for debugging. This happened when I was debugging some of the current bip test code we have and we were using a unicode string that was actually two bytes and then the first byte caused an error and was printed as a char like you suggest, took me ages to work out what was going on.

tcharding replied:
With the raw invalid byte value at least the dev debugging can think straight away "where did that byte come from" instead of wondering where that weird utf-8 char came from and having to convert it to a byte himself - at least that is what I had to do.

apoelstra replied:
So, I think we should be iterating over `hrp.chars()` in this loop, and then if we get a *char* that's out of bounds, we can stick that in the error. After checking it's in bounds we can safely cast it to a u8.

apoelstra replied:
In 179868360b76d5d9a8d806dd5721d51d84e84b48:

This is fine but I kinda think we should just fold this into `NonAsciiChar` (and rename this to `InvalidChar`). I don't think the user cares whether an invalid character was in the ASCII range or not.

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,420 @@
+            (true, false) => Case::Lower,
+            (false, true) => Case::Upper,
+            (false, false) => Case::None,
+            (_, _) => unreachable!("checked above"),
```
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

I think it would be nicer to just move the error return here. It means you lose the short-circuiting but it also means that you eliminate a panic branch.

tcharding replied:
How about this so we keep the short-circuiting
```
            (_, _) => return Err(MixedCase), // Defensive programming, this statement is unreachable.
```

apoelstra replied:
Yeah, I'd be cool with that.

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,420 @@
+use core::cmp::Ordering;
+use core::fmt::{self, Write};
+use core::iter::FusedIterator;
+use core::slice::Iter;
```
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

I think it's confusing to import this type. I'd import `slice` and refer to it as `slice::Iter` instead.

tcharding replied:
Ah yes, that is nicer.

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,420 @@
+pub struct ByteIter<'b> {
+    iter: Iter<'b, u8>,
+    size: usize, // Total size of the iterator.
+    n: usize,    // Number of characters yielded so far.
```
In 599ce6d089904386ac6a34d77a6b3b1fe222393f:

Why do we need to keep track of this stuff? Why can't `slice::Iter` do it for us?

tcharding replied:
Ha, overthought that didn't I.

tcharding replied:
Damn, I should trust myself more when reading reviews, we keep track of this because if we use `slice::Iter` it includes all the zeros in the hrp internal buffer.

apoelstra replied:
Then you should create the `slice::Iter` from a slice of the correct size :)

tcharding replied:
oooo, I should trust reviewers more ... :)

tcharding commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,420 @@
+    fn next(&mut self) -> Option<char> {
+        // Usage of unsafe is ok because we uphold the invariant that all
+        // bytes in the HRP are valid ASCII values.
+        self.iter.next().map(|b| unsafe { char::from_u32_unchecked(b as u32) })
```
Sweeeeet.

apoelstra replied:
Yeah :) I've spent a lot of time grepping the docs for `char` for `Into` and `From` impls for this crate..

tcharding commented:
Worked out why I couldn't get `duplicate_crash` to work - I didn't realise the output from `./fuzz.sh` over multiple lines was actually a single line if input, I was treating it as multiple lines of input. (I don't remember right now where we were discussing this @apoelstra.)

tcharding commented:
Changes in force push:
- A bunch of fuzzing fixes including adding a new fuzz target
- Iterate over chars as discussed above

tcharding commented:
Changes in force push:
- Used `iter::Take<slice::Iter<'b, u8>>` for the internal byte iterator so it has the correct length
- Improved the rustdocs to be more uniform
- Simplified the parsing code a little with no change of logic (removed early return since it is an optimisation and makes the code marginally more difficult to read)

apoelstra commented on fuzz/fuzz_targets/parse_hrp.rs:
```diff
@@ -0,0 +1,59 @@
+
+    // Make sure parsing garbage doesn't make us crash (from_utf8_lossy should
+    // contain some garbage, perhaps even invalid chars).
+    let _ = Hrp::parse(&s);
```
`from_utf8_lossy` will not include any invalid characters, at least not in the UTF8 sense.

I think we should use [str::from_utf8[(https://doc.rust-lang.org/std/str/fn.from_utf8.html) and return on error. This avoids an allocation and avoids mutating the input string, so it should be faster.

tcharding replied:
oooo, it replaces invalid chars with `?` - that's where those all came from when I tried to iterate over all the zeros in the buffer :)

> Converts a slice of bytes to a string, including invalid characters.

ref: https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy

I think your right, I'll try with `from_utf`. Thanks

apoelstra commented on fuzz/fuzz_targets/encode_decode.rs:
```diff
@@ -13,26 +15,22 @@ fn do_test(data: &[u8]) {
-    let hrp = String::from_utf8_lossy(&data[1..hrp_end])
-        .to_lowercase()
-        .to_string();
+    let hrp = String::from_utf8_lossy(&data[1..hrp_end]).to_lowercase().to_string();
```
In 179868360b76d5d9a8d806dd5721d51d84e84b48:

Did rustfmt make all these changes to this fuzz unit? They're a little distracting.

I have a bunch of comments about this fuzztest but since it already exists it's probably better to deal with it in a separate PR..

tcharding replied:
Sorry about that, I removed the rustfmt changes. PR still has changes to this fuzz test, I took your comments on the other one and applied it to this test. We have to touch this fuzz test in the PR because this PR changes the API.

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,406 @@
+            if !c.is_ascii() {
+                return Err(NonAsciiChar(c));
+            }
+            let b: u8 = c.try_into().expect("cannot fail for ASCII values");
```
In 179868360b76d5d9a8d806dd5721d51d84e84b48:

I think you should just write `let b = c as u8; // cast OK as we just checked that `c` is an ASCII value` which is shorter and eliminates a branch, a panic, and a static string, none of which I really trust the compiler to get rid of.

tcharding replied:
Sure thing, thanks for the reasoning.

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,406 @@
+    fn len(&self) -> usize { self.iter.len() }
+}
+
+impl<'b> FusedIterator for ByteIter<'b> {}
```
In 179868360b76d5d9a8d806dd5721d51d84e84b48:

I guess we ought to implement `DoubleEndedIterator` as well for good measure.

tcharding replied:
Done for each iterator and tested (noticed I missed on FusedIterator impl also).

apoelstra commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,406 @@
+    /// > specific applications.
+    ///
+    /// [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+    pub fn parse(hrp: &str) -> Result<Self, Error> { Ok(Self::parse_and_case(hrp)?.0) }
```
In 179868360b76d5d9a8d806dd5721d51d84e84b48:

Would be nice to add a `parse_unchecked` function that was `const`. It could call `str::as_bytes` and just copy data over. We don't even need to mark it `unsafe` because no single-byte value is an invalid character (or at least, `char::from<u8>()` somehow deals with the ones that are).

Would save people (and us) a lot of `parse("obviouslycorrect").unwrap()` code.

apoelstra replied:
Or if you wanted to preserve invariants we could just check everything is in the 33..126 range and replace any invalid bytes (or the whole string) with `x` or something.

tcharding replied:
I just paniced if the input is invalid.

apoelstra replied:
Interesting, I didn't know you could panic in constfns. I guess you just can't `unwrap`?

apoelstra replied:
I dunno if `unchecked` is the right word now for this :P given that the input is, in fact, checked. But I think it's fine. It signals to the user that they've gotta do their own validation.

apoelstra replied:
So, `panic` is not actually const in 1.48. I found a bug in my local CI setup that was causing the set of compilers to be sometimes collapsed into one. But I don't know why the github CI didn't catch this, and isn't catching this.

`cargo +1.48.0 test` fails right now on master.

apoelstra commented:
Done reviewing 179868360b76d5d9a8d806dd5721d51d84e84b48. Looking a lot better!

tcharding commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,406 @@
+
+impl<'b> Iterator for ByteIter<'b> {
+    type Item = u8;
+    fn next(&mut self) -> Option<u8> { self.iter.next().cloned() }
```
Done!

tcharding commented on src/primitives/hrp.rs:
```diff
@@ -0,0 +1,406 @@
+    ///
+    /// If an uppercase HRP was parsed during object construction then this iterator will yield
+    /// uppercase ASCII `char`s. For lowercase bytes see [`Self::lowercase_byte_iter`]
+    pub fn byte_iter(&self) -> ByteIter { ByteIter { iter: self.buf.iter().take(self.size) } }
```
Used as suggested, thanks. We'll get there in the end ;)

tcharding commented:
Thanks for your patience with so many review cycles @apoelstra!

apoelstra reviewed (approved):
ACK e8a1a5ec6915c87f7cc599a6c47bc000ebf70e3e

apoelstra commented:
@clarkmoody I'm happy with this PR now. ok to merge?

clarkmoody reviewed (approved):
This is really good stuff 👌 

Just a couple wishlist items that don't affect my approval:

- A new function `Hrp::empty()` or `impl Default` that replaces the two `Hrp { buf: [0_u8; MAX_HRP_LEN], size: 0 };` statements.
- (Is this out of scope for this library?) Some constants for Bitcoin HRPs:

```rust
impl Hrp {
    const BITCOIN: Self = Self { buf: [b'b', b'c', 0, 0, 0...], size: 2 };
    // Testnet too
}
```

tcharding commented:
I added `default` and one usage of it (we cannot call it in the const function).

Can we leave the consts to another PR as we see how the separation between `rust-bech32` and `rust-bitcoin` develops?

apoelstra commented:
We can leave both to a future PR since they're small enough that they'll be quick to review (and not cause any conflicts).
